pysynchronizer/icbinn.py

```diff
@@ -0,0 +1,241 @@
+                                  "icbinn_rand failed: %s" % exc)
+        return data
+
+    def write_file(self, name, content):
+        file_obj = self.open(name, O_WRONLY | O_CREAT)
+        file_obj.pwrite(content, 0)
+        file_obj.close()
+
+    def mounted_path(self, path):
+        components = path.split('/')
+        if '.' in components or '..' in components:
+            raise TargetStateError('invalid components in %s' % (path))
+        return join(self.mount_point, path)
+
+class IcbinnFile(object):
+    def __init__(self, icbinn, path, mode):
+        self.icbinn = icbinn
+        self.path = path
+        self.write_offset = 0
+
+        self.fd = icbinn_open(self.icbinn, str(self.path), mode)
+        if self.fd < 0:
+            raise IcbinnError("error opening icbinn file '%s': icbinn_open "
+                              "failed" % self.path)
+
+    def close(self):
+        if icbinn_close(self.icbinn, self.fd) < 0:
+            raise IcbinnError("error closing icbinn file '%s': icbinn_close "
+                              "failed" % self.path)
+
+    def get_read_lock(self):
+        if icbinn_lock(self.icbinn, self.fd, ICBINN_LTYPE_RDLCK) < 0:
+            raise IcbinnError("error getting read lock on icbinn file '%s': "
+                              "icbinn_lock failed" % self.path)
+
+    def get_write_lock(self):
+        if icbinn_lock(self.icbinn, self.fd, ICBINN_LTYPE_WRLCK) < 0:
+            raise IcbinnError("error getting write lock on icbinn file '%s': "
+                              "icbinn_lock failed" % self.path)
+
+    def seek(self, offset):
+        self.write_offset = offset
+
+    def write(self, data):
+        try:
+            self.pwrite(data, self.write_offset)
+        except IcbinnError as e:
+            raise e
+
+        self.write_offset += len(data)
+
+    def pwrite(self, data, offset):
+        data_len = len(data)
+        written = 0
+        while written < data_len:
+            n = icbinn_pwrite(self.icbinn, self.fd,
+                              data[written:written + ICBINN_MAXDATA],
+                              offset + written)
+            if n < 0:
+                raise IcbinnError("error writing to icbinn file '%s': "
+                                  "icbinn_pwrite failed" % self.path)
+            written += n
+
+    def pread(self, size, offset):
+        try:
+            return icbinn_pread(self.icbinn, self.fd, size, offset)
+        except IOError as exc:
+            raise IcbinnError("error reading icbinn file '%s': icbinn_pwrite "
+                              "failed: %s" % (self.path, exc))
+
+    def unlock(self):
+        if icbinn_lock(self.icbinn, self.fd, ICBINN_LTYPE_UNLCK) < 0:
+            raise IcbinnError("error unlocking icbinn file '%s': icbinn_lock "
+                              "failed" % self.path)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *_):
```

Comment:
According to the [python documentation](https://docs.python.org/3/reference/datamodel.html#object.__exit__) the prototype for  __exit__ is,
```
object.__exit__(self, exc_type, exc_value, traceback)
```
The `*_` notation allows for dropping the last three args without having to declare three unused variables. If the consensus is to not use `*_`, then it would have to be replaced with three unused variables otherwise a runtime exception will occur when `with` is used.

Reply:
Yep, you're right, I completely missed that.

Reply:
No worries. Python docs do that a lot, they will ref a function with empty parens and you have to trace down the actual def to see what the prototype actually is and it won't bite you until the actual call is made at execution time.